activation/wire/wire_v2.go

```diff
@@ -42,20 +46,96 @@ type ActivationTxV2 struct {
 
 	SmesherID types.NodeID
 	Signature types.EdSignature
+
+	// cached fields to avoid repeated calculations
+	id types.ATXID
 }
 
-// TODO: This is dummy implementation for testing.
-// It needs to be implemented properly.
 func (atx *ActivationTxV2) SignedBytes() []byte {
-	atxCopy := *atx
-	atxCopy.Signature = types.EdSignature{}
-	return codec.MustEncode(&atxCopy)
+	return atx.ID().Bytes()
 }
 
-// TODO: This is dummy implementation for testing.
-// It needs to be implemented properly.
 func (atx *ActivationTxV2) ID() types.ATXID {
-	return types.BytesToATXID(atx.SignedBytes())
+	if atx.id != types.EmptyATXID {
+		return atx.id
+	}
+
+	tree, err := merkle.NewTreeBuilder().
+		WithHashFunc(atxTreeHash).
+		Build()
+	if err != nil {
+		panic(err)
+	}
+
+	tree.AddLeaf((*[4]byte)(unsafe.Pointer(&atx.PublishEpoch))[:])
+	tree.AddLeaf(atx.PositioningATX.Bytes())
+	if atx.Coinbase != nil {
+		tree.AddLeaf(atx.Coinbase.Bytes())
+	} else {
+		tree.AddLeaf(types.Address{}.Bytes())
+	}
+	if atx.Initial != nil {
+		tree.AddLeaf(atx.Initial.Root())
+	} else {
+		tree.AddLeaf(types.EmptyHash32.Bytes())
+	}
+
+	prevATXTree, err := merkle.NewTreeBuilder().
+		WithHashFunc(atxTreeHash).
+		Build()
+	if err != nil {
+		panic(err)
+	}
+	for _, prevATX := range atx.PreviousATXs {
+		prevATXTree.AddLeaf(prevATX.Bytes())
+	}
+	for i := 0; i < 256-len(atx.PreviousATXs); i++ {
```

Comment:
NIT, it's simpler to reason about:
```suggestion
	for i := len(atx.PreviousATXs); i < 256; i++ {
```